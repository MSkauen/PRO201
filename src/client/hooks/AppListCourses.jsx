import React, { useEffect } from "react";
import { LoadingView } from "../components/LoadingView";
import { useLoading } from "../lib/useLoading";
import { ErrorView } from "../components/ErrorView";
import "../../shared/css/stylesheet.css";
import "../../shared/css/chooseCourse.css";
import { PRODUCTS } from "../lib/images.jsx"
import { fetchJson } from "../lib/http";
import { Link } from "react-router-dom";
import { useParams } from "react-router";
import { closeModal } from "../components/ModalView";
import lock from "url:../../shared/img/locked.png";

export function AppListCourses({user}) {

  useEffect(() => {
    window.onclick = function(event) {
      const modal = document.getElementById("myModal")

      if (event.target === modal) {
        modal.style.display = "none";
      }
    }
  })

  const { data, error, loading, reload } = useLoading(() =>
      fetchJson("/api/profile", {
        method: "POST",
        body: JSON.stringify({}),
        headers: {
          "Content-Type": "application/json",
        },
      })
  );

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }
  if (loading || !data) {
    return <LoadingView />;
  }


  return (
    <>

      <div id="myModal" className="modal">
        <div className="modal-content">
          <span onClick={closeModal} className="close">x</span>
          <p>
            Please enter your username in the field below.
          </p>
        </div>
      </div>

      <div className="chooseCourseContainer">
        <div className="productsContainer">
          {
            user.courses.map((id) => (
                  id.access !== true
                      ?
                        <div key={id.id} tabIndex="0" className="bigDot">
                          <img id="lock" src={lock} alt=""/>
                        </div>
                      :
                      <Link key={id.id} to={`/courses/${user.username}/watch/${id.id}/0`}>
                        <div tabIndex="0" className="bigDot">
                          <img src={PRODUCTS[id.id].image} alt=""/>
                        </div>
                      </Link>
            ))}
        </div>
      </div>
    </>
  );
}

export function GetCourses({ userApi }) {
  const { id } = useParams()
  const { data: user, loading, error, reload } = useLoading(
      async () => await userApi.getUserData(id),
      [id]
  );
  if (error) {
    return <ErrorView error={error} reload={reload()} />;
  }

  if (loading || !user) {
    return <LoadingView />;
  }

  return <AppListCourses user={user} />;
}