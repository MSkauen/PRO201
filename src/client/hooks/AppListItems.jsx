import React from "react";
import { LoadingView } from "../components/LoadingView";
import { useLoading } from "../lib/useLoading";
import { ErrorView } from "../components/ErrorView";
import { BrowserRouter, Link } from "react-router-dom";
import {useParams} from "react-router";
import {LogPage} from "./LogPage";


export function AppListItems({item}) {

  return (
    <>

      <h1>PARTS ${item.partsChanged } REPLACED FOR ${item.id}</h1>
      <div className="partsReplacedContainer">
      </div>

      <div className="success-checkmark">
        <div className="check-icon">
          <span className="icon-line line-tip"></span>
          <span className="icon-line line-long"></span>
          <div className="icon-circle"></div>
          <div className="icon-fix"></div>
        </div>
      </div>

      <h1>SUCCESS!</h1>
      <h2 className="timerTag"></h2>

      <h1>List messages</h1>

    </>
  );
}
export function GetItem({ itemApi }) {
    const { id } = useParams()
    const { data: item, loading, error, reload } = useLoading(
        async () => await itemApi.getItem(id),
        [id]
    );
    if (error) {
        return <ErrorView error={error} reload={reload()} />;
    }

    if (loading || !item) {
        return <LoadingView />;
    }

    return <AppListItems item={item} />;
}

//      {messages.map(({ id, sender, subject }) => (
//         <li key={id}>
//           <Link to={`/messages/${id}/edit`}>
//             <b>{sender}</b> [Subject:{subject}]
//           </Link>
//         </li>
//       ))}