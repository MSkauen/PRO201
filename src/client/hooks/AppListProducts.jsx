import React from "react";
import { LoadingView } from "../components/LoadingView";
import { useLoading } from "../lib/useLoading";
import { ErrorView } from "../components/ErrorView";
import { Link } from "react-router-dom";
import "../../shared/css/stylesheet.css";
import { PRODUCTS } from "../lib/images.jsx"
import {fetchJson} from "../lib/http";

export function AppListProducts() {

  const { data, error, loading, reload } = useLoading(() =>
      fetchJson("/api/profile", {
        method: "POST",
        body: JSON.stringify({}),
        headers: {
          "Content-Type": "application/json",
        },
      })
  );

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }
  if (loading || !data) {
    return <LoadingView />;
  }

  return (
    <>
      <div className="chooseCourseContainer">
        <div className="productsContainer">
          <a href="https://mskauen.github.io/pro201eksamen/sunbellCourse.html">
            <div tabIndex="0" className="bigDot">
              <img src={PRODUCTS[0].image} alt=""/>
            </div>
          </a>
          <a href="https://mskauen.github.io/pro201eksamen/moveSmartCourse.html">
            <div tabIndex="0" className="bigDot">
              <img src="img/products/Bright_Move_Red.png" alt=""/>
                <div/>
            </div>
          </a>
          <a href="https://mskauen.github.io/pro201eksamen/sunbellCourse.html">
            <div tabIndex="0" className="bigDot">
              <img src="img/products/sunturtle-red.png" alt=""/>
                <div/>
            </div>
          </a>
        </div>
      </div>
      <h1>List users</h1>
      {users.map(({ id, firstName, lastName }) => (
        <li key={id}>
          <Link to={`/users/${id}/edit`}>
            {firstName} {lastName}
          </Link>
        </li>
      ))}
    </>
  );
}
