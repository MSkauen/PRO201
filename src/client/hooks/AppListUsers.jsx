import React from "react";
import { LoadingView } from "../components/LoadingView";
import { useLoading } from "../lib/useLoading";
import { ErrorView } from "../components/ErrorView";
import { Link } from "react-router-dom";
import "../../shared/css/stylesheet.css";

export function AppListUsers({ userApi }) {
  const { data: users, error, loading, reload } = useLoading(
    async () => await userApi.listUsers()
  );

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }
  if (loading || !users) {
    return <LoadingView />;
  }

  return (
    <>
      <h1>List users</h1>
      {users.map(({ id, firstName, lastName }) => (
        <li key={id}>
          <Link to={`/users/${id}/edit`}>
            {firstName} {lastName}
          </Link>
        </li>
      ))}
    </>
  );
}
