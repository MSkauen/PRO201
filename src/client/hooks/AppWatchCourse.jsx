import React, { useEffect } from "react";
import { LoadingView } from "../components/LoadingView";
import { useLoading } from "../lib/useLoading";
import { ErrorView } from "../components/ErrorView";
import { fetchJson } from "../lib/http";
import "../../shared/css/main.css";
import {useHistory, useParams} from "react-router";
import lock from "url:../../shared/img/locked.png";
import { MISC } from "../lib/images.jsx"
import Youtube from "react-youtube"
import {closeModal} from "../components/ModalView";
import {Link} from "react-router-dom";

export function AppWatchCourse({user, courseId, coursePartId}) {
    const history = useHistory();

    let currentCourseName = user.courses[courseId].courseParts[coursePartId].name
    let trimmedCourseName = currentCourseName.replace(/[^a-zA-Z ]/g, "")

    let videoUrl = user.courses[courseId].courseParts[coursePartId].contentUrl
    let videoCode

    if (videoUrl) {
        videoCode = videoUrl.split("embed/")[1].split("&")[0]
    }

    useEffect(() => {
        window.onclick = function(event) {
            const modal = document.getElementById("myModal")

            if (event.target === modal) {
                modal.style.display = "none";
            }
        }
    })

    const { data, error, loading, reload } = useLoading(() =>
        fetchJson("/api/profile", {
            method: "POST",
            body: JSON.stringify({}),
            headers: {
                "Content-Type": "application/json",
            },
        })
    );

    const checkElapsedTime = (e) => {
        console.log(e.target.getCurrentTime());
        const duration = e.target.getDuration();
        const currentTime = e.target.getCurrentTime();
        if (currentTime / duration > 0.95) {
            console.log("completed")
        }
    };

    if (error) {
        return <ErrorView error={error} reload={reload} />;
    }
    if (loading || !data) {
        return <LoadingView />;
    }

    return (
    <>
        <div id="backButton" onClick={()=> (history.push(`/courses/${user.username}`))}/>

        <div id="myModal" className="modal">
            <div className="modal-content">
                <span onClick={closeModal} className="close">x</span>
                <p>
                    Please enter your username in the field below.
                </p>
            </div>
        </div>

        <div className="courseContainer">
            <div className="videoContainer">
                <h1>{trimmedCourseName}</h1>
                <Youtube className="youtubePlayer" videoId={videoCode} onStateChange={(e) => checkElapsedTime(e)}
                />
            </div>
            <div className="sideBar">

                <div className="course">
                    <div className="courseDetails">
                        <h2>{user.courses[courseId].name}</h2>
                        <h2>{(parseInt(coursePartId) + 1)}/{user.courses[courseId].courseParts.length}</h2>
                    </div>
                    <div className="allVideosContainer">

                        {
                            user.courses[courseId].courseParts.map((id) => (
                                <div key={id.id} className="sideBarItem">
                                    <h4>{id.name}</h4>
                                    <div className="playButton">
                                        {
                                            id.access !== true
                                                ?
                                                <img id="lock" className="lockedIcon" src={lock} alt=""/>
                                                :
                                                <Link key={id.id} to={`/courses/${user.username}/watch/${courseId}/${id.id}`}>
                                                    <img className="playIcon" src={MISC[1].image} alt=""/>
                                                </Link>
                                        }
                                    </div>
                                </div>
                            ))}

                    </div>
                </div>
            </div>
        </div>
    </>
    );
}

export function GetUser({ userApi }) {
    const { id } = useParams()
    const { courseId } = useParams()
    const { coursePartId } = useParams()

    const { data: user, loading, error, reload } = useLoading(
        async () => await userApi.getUserData(id),
        [id]
    );
    if (error) {
        return <ErrorView error={error} reload={reload()} />;
    }

    if (loading || !user) {
        return <LoadingView />;
    }

    return <AppWatchCourse user={user} courseId={courseId} coursePartId={coursePartId} />;
}
