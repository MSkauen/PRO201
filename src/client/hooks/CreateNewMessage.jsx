import React, { useState } from "react";
import { InputField } from "../components/InputField";
import "../../shared/css/stylesheet.css";

export function CreateNewMessage({ messageApi }) {
  const [subject, setSubject] = useState("");
  const [recipient, setRecipient] = useState("");
  const [content, setContent] = useState("");
  const [date, setDate] = useState("");

  async function submit(e) {
    e.preventDefault();

    const res = await fetch("/api/profile", {});
    const json = await res.json();
    let sender = json.username;

    console.log("Submitting", { sender, recipient, subject, content, date });
    await fetch("/api/messages", {
      method: "POST",
      body: JSON.stringify({ sender, recipient, subject, content, date }),
      headers: {
        "Content-Type": "application/json",
      },
    });
  }

  return (
    <form onSubmit={submit}>
      <h1>Create a new message</h1>
      <InputField
        label={"Subject"}
        value={subject}
        onChangeValue={setSubject}
      />
      <InputField
        label={"Recipient"}
        value={recipient}
        onChangeValue={setRecipient}
      />
      <InputField
        label={"Content"}
        value={content}
        onChangeValue={setContent}
      />
      <InputField
        label={"Date"}
        value={date}
        onChangeValue={setDate}
        type="date"
      />
      <button>Submit</button>
    </form>
  );
}
