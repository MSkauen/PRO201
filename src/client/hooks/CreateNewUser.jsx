import React, { useState } from "react";
import { InputField } from "../components/InputField";
import "../../shared/css/stylesheet.css";

export function CreateNewUser({ messapeApi }) {
  const [firstName, setFirstName] = useState("");
  const [lastName, setLastName] = useState("");
  const [email, setEmail] = useState("");

  async function submit(e) {
    e.preventDefault();
    console.log("Submitting", { firstName, lastName, email });
    await fetch("/api/users", {
      method: "POST",
      body: JSON.stringify({ firstName, lastName, email }),
      headers: {
        "Content-Type": "application/json",
      },
    });
  }

  return (
    <form onSubmit={submit}>
      <h1>Create a new user</h1>
      <InputField
        label={"First name"}
        value={firstName}
        onChangeValue={setFirstName}
      />
      <InputField
        label={"Last name"}
        value={lastName}
        onChangeValue={setLastName}
      />
      <InputField label={"Email"} value={email} onChangeValue={setEmail} />
      <button>Submit</button>
    </form>
  );
}
