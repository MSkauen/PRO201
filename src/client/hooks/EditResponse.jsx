import React, { useState } from "react";
import { useParams } from "react-router";
import { LoadingView } from "../components/LoadingView";
import { InputField } from "../components/InputField";
import { ErrorView } from "../components/ErrorView";
import { useLoading } from "../lib/useLoading";
import "../../shared/css/stylesheet.css";

export function EditMessageForm({ message }) {
  const [subject, setSubject] = useState(message.subject);
  const [recipient, setRecipient] = useState(message.sender);
  const [content, setContent] = useState();
  const [date, setDate] = useState(message.date);
  let sender;

  async function submit(e) {
    e.preventDefault();

    const res = await fetch("/api/profile", {});
    const json = await res.json();
    sender = json.username;

    console.log("Submitting", { sender, recipient, subject, content, date });
    await fetch(`/api/messages/`, {
      method: "POST",
      body: JSON.stringify({ sender, recipient, subject, content, date }),
      headers: {
        "Content-Type": "application/json",
      },
    });
  }

  return (
    <form onSubmit={submit}>
      <h1>{subject}</h1>
      <p>"{message.content}"</p>
      <InputField
        label={"Recipient"}
        value={recipient}
        onChangeValue={setRecipient}
      />
      <InputField
        label={"Subject"}
        value={subject}
        onChangeValue={setSubject}
      />
      <InputField
        label={"Content"}
        value={content}
        onChangeValue={setContent}
      />
      <InputField
        label={"Date"}
        value={date}
        onChangeValue={setDate}
        type="date"
      />
      <button>Submit</button>
    </form>
  );
}

export function EditResponse({ messageApi }) {
  const { id } = useParams();

  const { data: message, loading, error, reload } = useLoading(
    async () => await messageApi.getMessage(id),
    [id]
  );

  if (error) {
    return <ErrorView error={error} reload={reload()} />;
  }

  if (loading || !message) {
    return <LoadingView />;
  }

  return <EditMessageForm message={message} />;
}
