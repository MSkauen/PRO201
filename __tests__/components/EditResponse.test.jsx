import { EditMessageForm } from "../../src/client/hooks/EditResponse";

import * as ReactDOM from "react-dom";
import * as React from "react";
import { MemoryRouter } from "react-router";
import { act } from "react-dom/test-utils";

const message = {
  listMessages: async () => [
    {
      id: 1,
      subject: "Stuff",
      recipient: "admin",
      content: "Hello test",
      date: "",
    },
  ],
};

describe("Response view", () => {
  it("show response subject on dom", async () => {
    const container = document.createElement("div");
    document.body.appendChild(container);
    await act(async () => {
      ReactDOM.render(
        <MemoryRouter>
          <EditMessageForm message={message} />
        </MemoryRouter>,
        container
      );
    });

    expect(container.innerHTML).toMatchSnapshot();
    expect(container.querySelector("input").textContent).toEqual("");
  });
});
