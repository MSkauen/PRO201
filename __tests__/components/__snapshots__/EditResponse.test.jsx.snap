// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Response view show response subject on dom 1`] = `"<form><h1></h1><p>\\"\\"</p><div><label>Recipient: <input value=\\"\\"></label></div><div><label>Subject: <input value=\\"\\"></label></div><div><label>Content: <input value=\\"\\"></label></div><div><label>Date: <input type=\\"date\\" value=\\"\\"></label></div><button>Submit</button></form>"`;
